File: Common/SaveDelayer.vb
Public Class SaveDelayer

    Public Sub New(ByVal Stream As IO.Stream, Optional ByVal Encoding As Text.Encoding = Nothing)
        If Encoding Is Nothing Then
            Encoding = Text.Encoding.UTF8
        End If

        Me._Stream = Stream
        Me._Encoding = Encoding
    End Sub

    Private Sub DoSave()
        Me.Stream.Position = 0
        Dim N = Me.Stream.Write(Me.InStream, Me.Length)
        Me.Stream.SetLength(N)
    End Sub

    Public Sub Save(ByVal Stream As IO.Stream, ByVal Length As Integer, ByVal RunningMode As TaskDelayerRunningMode)
        Me.InStream = Stream
        Me.Length = Length
        Me.TaskDelayer.RunTask(RunningMode)
    End Sub

    Public Sub Save(ByVal Stream As IO.Stream, ByVal RunningMode As TaskDelayerRunningMode)
        Me.Save(Stream, -1, RunningMode)
    End Sub

    Public Sub Save(ByVal Buffer As Byte(), ByVal Index As Integer, ByVal Length As Integer, ByVal RunningMode As TaskDelayerRunningMode)
        Dim Strm = New IO.MemoryStream(Buffer, Index, Length)
        Me.Save(Strm, RunningMode)
    End Sub

    Public Sub Save(ByVal Buffer As Byte(), ByVal RunningMode As TaskDelayerRunningMode)
        Me.Save(Buffer, 0, Buffer.Length, RunningMode)
    End Sub

    Public Sub Save(ByVal Str As String, ByVal RunningMode As TaskDelayerRunningMode)
        Me.Save(Me.Encoding.GetBytes(Str), RunningMode)
    End Sub

#Region "Stream Property"
    Private ReadOnly _Stream As IO.Stream

    Public ReadOnly Property Stream As IO.Stream
        Get
            Return Me._Stream
        End Get
    End Property
#End Region

#Region "Encoding Property"
    Private ReadOnly _Encoding As Text.Encoding

    Public ReadOnly Property Encoding As Text.Encoding
        Get
            Return Me._Encoding
        End Get
    End Property
#End Region

    Private ReadOnly TaskDelayer As TaskDelayer = New TaskDelayer(AddressOf Me.DoSave, 10000)
    Private InStream As IO.Stream
    Private Length As Integer

End Class

File: Common/SaveDelayer.cs
using System;

namespace Ks.Common
{
    public class SaveDelayer
    {

        // ToDo Implement IDisposable support.

        public SaveDelayer(System.IO.Stream Stream, System.Text.Encoding Encoding = null)
        {
            this.TaskDelayer = new TaskDelayer(this.DoSave, TimeSpan.FromSeconds(10));

            if (Encoding == null)
            {
                Encoding = System.Text.Encoding.UTF8;
            }

            this.Stream = Stream;
            this.Encoding = Encoding;
        }

        private void DoSave()
        {
            this.Stream.Position = 0;
            var N = this.Stream.Write(this.InStream, this.Length);
            this.Stream.SetLength(N);
        }

        public void Save(System.IO.Stream Stream, int Length, TaskDelayerRunningMode RunningMode)
        {
            this.InStream = Stream;
            this.Length = Length;
            this.TaskDelayer.RunTask(RunningMode);
        }

        public void Save(System.IO.Stream Stream, TaskDelayerRunningMode RunningMode)
        {
            this.Save(Stream, -1, RunningMode);
        }

        public void Save(byte[] Buffer, int Index, int Length, TaskDelayerRunningMode RunningMode)
        {
            var Strm = new System.IO.MemoryStream(Buffer, Index, Length);
            this.Save(Strm, RunningMode);
        }

        public void Save(byte[] Buffer, TaskDelayerRunningMode RunningMode)
        {
            this.Save(Buffer, 0, Buffer.Length, RunningMode);
        }

        public void Save(string Str, TaskDelayerRunningMode RunningMode)
        {
            this.Save(this.Encoding.GetBytes(Str), RunningMode);
        }

        public System.IO.Stream Stream { get; }

        public System.Text.Encoding Encoding { get; }

        private readonly TaskDelayer TaskDelayer;
        private System.IO.Stream InStream;
        private int Length;
    }
}
